File: src/structured/binary_search.c
#include "binary_search.h"

Result binarySearch(int arr[], int arrSize, int value) {
  Result result;

  int low = 0;
  int high = arrSize - 1;

  while (low <= high) {
    int middle = (low + high) / 2;
    int guess = arr[middle];

    if (guess == value) {
      result.index = middle;
      result.value = value;
      return result;
    }

    if (guess > value) {
      high = middle - 1;
    } else {
      low = middle + 1;
    }
  }

  result.index = -1;
  result.value = -1;
  return result;
}

File: Go/internal/recursive/binary_search.go
package recursive

func BinarySearch(arr []int, val int) int {
	if arr == nil || len(arr) == 0 {
		return -1
	}
	lowestIndex := 0
	highestIndex := len(arr) - 1
	middleIndex := (highestIndex + lowestIndex) / 2
	guess := arr[middleIndex]
	if guess == val {
		return middleIndex
	}
	if guess > val {
		highestIndex = middleIndex - 1
		return BinarySearch(arr[:highestIndex], val)
	}
	if guess < val {
		lowestIndex = middleIndex + 1
		return BinarySearch(arr[lowestIndex:], val)
	}
	return -1
}
